### PR TITLE
feat: add retention analysis api

### PR DESCRIPTION
**Pull request by llmin**
----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

## Summary

(describe what this merge request does)

## Implementation highlights

(describe how the merge request does for feature changes, share the RFC link if it has)

## Test checklist

- [x] add new test cases
- [ ] all code changes are covered by unit tests
- [ ] end-to-end tests
  - [ ] deploy web console with CloudFront + S3 + API gateway
  - [ ] deploy web console within VPC
  - [ ] deploy ingestion server
    - [ ] with MSK sink
    - [ ] with KDS sink
    - [ ] with S3 sink
  - [ ] deploy data processing
  - [ ] deploy data modeling
    - [ ] new Redshift Serverless
    - [ ] provisioned Redshift
    - [ ] Athena
  - [ ] deploy with reporting

## Is it a breaking change

- [ ] add parameters without default value in stack
- [ ] introduce new service permission in stack
- [ ] introduce new top level stack module

## Miscellaneous

- [ ] introduce new symbol link source file(s) to be shared among infra code, web console frontend, and web console backend

### CODE REVIEW

File: src/control-plane/backend/lambda/api/test/api/reporting.test.ts [hunk 1]
```diff
@@ -614,6 +614,106 @@ describe('reporting test', () => {
...
+        pipelineId: 'pipeline-1111111',
+        appId: 'app1',
+        sheetName: 'sheet99',
+        computeMethod: 'USER_CNT',
+        specifyJoinColumn: true,
+        joinColumn: 'user_pseudo_id',
+        conversionIntervalType: 'CUSTOMIZE',
+        conversionIntervalInSeconds: 7200,
+        eventAndConditions: [{
+          eventName: 'add_button_click',
```
Comment (tyyzqmf) [inline src/control-plane/backend/lambda/api/test/api/reporting.test.ts:657]: add test case with some conditions

File: src/control-plane/backend/lambda/api/test/api/sql-builder.test.ts [hunk 2]
```diff
@@ -3533,5 +3533,259 @@ describe('SQL Builder test', () => {
...
+    const sql = buildRetentionAnalysisView('app1', 'testview', {
+      schemaName: 'app1',
+      computeMethod: ExploreComputeMethod.USER_CNT,
+      specifyJoinColumn: true,
+      joinColumn: 'user_pseudo_id',
+      conversionIntervalType: ExploreConversionIntervalType.CUSTOMIZE,
+      conversionIntervalInSeconds: 10*60,
+      eventAndConditions: [
+        {
+          eventName: 'add_button_click',
```
Comment (tyyzqmf) [inline src/control-plane/backend/lambda/api/test/api/sql-builder.test.ts:3547]: add test case with some conditions

File: src/control-plane/backend/lambda/api/test/api/reporting.test.ts (hunk 1 above)
Comment (llmin) [reply src/control-plane/backend/lambda/api/test/api/reporting.test.ts:657]: The conditions have already been covered in other test cases, adding more conditions in this test case won't have any effect. It only affects the SQL statement, which is already covered in sql-builder.test.ts.

File: src/control-plane/backend/lambda/api/service/reporting.ts [hunk 3]
```diff
@@ -238,8 +241,8 @@ export class ReportingServ {
         conversionIntervalInSeconds: query.conversionIntervalInSeconds,
         eventAndConditions: query.eventAndConditions,
         timeScopeType: query.timeScopeType,
-        timeStart: query.timeScopeType === ExploreTimeScopeType.FIXED ? query.timeStart : undefined,
-        timeEnd: query.timeScopeType === ExploreTimeScopeType.FIXED ? query.timeEnd : undefined,
+        timeStart: query.timeScopeType === 'FIXED' ? query.timeStart : undefined,
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/reporting.ts:244]: why rollback using enum value?

File: src/control-plane/backend/lambda/api/service/reporting.ts [hunk 4]
```diff
@@ -444,8 +447,115 @@ export class ReportingServ {
...
+      const sql = buildRetentionAnalysisView(query.appId, viewName, {
+        schemaName: query.appId,
+        computeMethod: query.computeMethod,
+        specifyJoinColumn: query.specifyJoinColumn,
+        joinColumn: query.joinColumn,
+        conversionIntervalType: query.conversionIntervalType,
+        conversionIntervalInSeconds: query.conversionIntervalInSeconds,
+        eventAndConditions: query.eventAndConditions,
+        timeScopeType: query.timeScopeType,
+        timeStart: query.timeScopeType === 'FIXED' ? query.timeStart : undefined,
```
Comment (zxkane) [inline src/control-plane/backend/lambda/api/service/reporting.ts:469]: use enum `ExploreTimeScopeType` instead?

**Review by zxkane — CHANGES_REQUESTED** [review]

File: src/control-plane/backend/lambda/api/service/reporting.ts (hunk 4 above)
Comment (llmin) [reply src/control-plane/backend/lambda/api/service/reporting.ts]: fixed

File: src/control-plane/backend/lambda/api/service/reporting.ts (hunk 3 above)
Comment (llmin) [reply src/control-plane/backend/lambda/api/service/reporting.ts]: fixed

**Review by zxkane — APPROVED** [review]

**Pull request merged**